File: card-lang-rust/src/calculator/parser/symbols/open-parenthesis.rs

use crate::types::Parser::Parser;
use crate::types::Parser::NudListener;
use crate::types::Parser::SymbolAndToken;
use crate::types::ReusedStructs::Node;

pub struct OpenParenthesisNud {}
impl NudListener for OpenParenthesisNud {
  fn run(&self, symtok: SymbolAndToken, parser: Parser)->Node{
    let (p, n) = parser.expression(2);
    let (p, t) = p.token();
    if t.token.token_type != ")" {
      panic!("Expected closing parenthesis ')'");
    }
    p.advance();
    return n;
  }
}

File: card-lang-typescript/src/calculator/parser/symbols/open-parenthesis.ts
import { NudListener, ParserController, SymbolAndToken } from "../../../types/Parser";
import { LangNode } from "../../../types/ReusedStructs";

export class OpenParenthesisNud implements NudListener {
  run(symtok: SymbolAndToken, parser: ParserController): LangNode{
    let n = parser.expression(2);
    let t = parser.token();
    if(typeof t == "undefined"){
      throw new Error(`Expected closing parenthesis ')', reached end of token list`);
    }
    if(t.token.values.value != ")"){
      throw new Error(`Expected closing parenthesis ')', got ${t.token.token_type}`);
    }
    parser.advance();
    return n;
  }
}
